crypto/s2n_certificate.c

```diff
@@ -176,30 +178,132 @@ struct s2n_cert_chain_and_key *s2n_cert_chain_and_key_new(void)
     memset(&chain_and_key->sct_list, 0, sizeof(chain_and_key->sct_list));
-    chain_and_key->san_names = NULL;
-    chain_and_key->x509_cert = NULL;
+    chain_and_key->cn_names = s2n_array_new(sizeof(struct s2n_blob));
+    if (!chain_and_key->cn_names) {
+        return NULL;
+    }
+
+    chain_and_key->san_names = s2n_array_new(sizeof(struct s2n_blob));
+    if (!chain_and_key->san_names) {
+        return NULL;
+    }
 
     return chain_and_key;
 }
 
-static int s2n_cert_chain_and_key_set_x509(struct s2n_cert_chain_and_key *chain_and_key, struct s2n_blob *leaf_bytes)
+int s2n_cert_chain_and_key_load_sans(struct s2n_cert_chain_and_key *chain_and_key, X509 *x509_cert)
 {
-    const unsigned char *leaf_der = leaf_bytes->data;
-    X509 *cert = d2i_X509(NULL, &leaf_der, leaf_bytes->size);
-    if (!cert) {
-        S2N_ERROR(S2N_ERR_INVALID_PEM);
-    }
+    notnull_check(chain_and_key->san_names);
 
-    chain_and_key->x509_cert = cert;
-
-    GENERAL_NAMES *san_names = X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL);
+    GENERAL_NAMES *san_names = X509_get_ext_d2i(x509_cert, NID_subject_alt_name, NULL, NULL);
     if (san_names == NULL) {
         /* No SAN extension */
         return 0;
     }
 
-    chain_and_key->san_names = san_names;
+    for (int i = 0; i < sk_GENERAL_NAME_num(san_names); i++) {
+        GENERAL_NAME *san_name = sk_GENERAL_NAME_value(san_names, i);
+        if (!san_name) {
+            continue;
+        }
 
+        if (san_name->type == GEN_DNS) {
+            /* Per https://www.openssl.org/docs/man1.1.0/man3/ASN1_STRING_data.html there may
+             * be embedded NULLs inside of the SAN string. The comparison later with any proper
+             * DNS hostnames will safely fail.
+             */
+            unsigned char *san_str = san_name->d.dNSName->data;
+            const size_t san_str_len = san_name->d.dNSName->length;
+            struct s2n_blob *san_blob = s2n_array_insert(chain_and_key->san_names, s2n_array_num_elements(chain_and_key->san_names));
+            if (!san_blob) {
+                GENERAL_NAMES_free(san_names);
+                return -1;
+            }
+
+            if (s2n_alloc(san_blob, san_str_len)) {
+                GENERAL_NAMES_free(san_names);
+                return -1;
+            }
+
+            memcpy_check(san_blob->data, san_str, san_str_len);
+            san_blob->size = san_str_len;
+        }
+    }
+
+    GENERAL_NAMES_free(san_names);
+    return 0;
+}
+
+int s2n_cert_chain_and_key_load_cn(struct s2n_cert_chain_and_key *chain_and_key, X509 *x509_cert)
+{
+    notnull_check(chain_and_key->cn_names);
+
+    X509_NAME *subject = X509_get_subject_name(x509_cert);
+    if (!subject) {
+        return 0;
+    }
+
+    /* Technically there can be multiple CommonNames entries in the Subject. */
+    int lastpos = -1;
+    while((lastpos = X509_NAME_get_index_by_NID(subject, NID_commonName, lastpos)) >= 0) {
```

Comment:
Where did the information about multiple CommonNames come from? I've been having a hard time finding a reference in the RFC or examples of having to parse multiple CNs. For example in the hostname validation example code on the openssl wiki, they parse a single CN but multiple SANs: https://wiki.openssl.org/index.php/Hostname_validation

Reply:
There isn't any wording I can find in https://www.rfc-editor.org/rfc/rfc5280.html#section-4.1.2.4 (4.1.2.4 defines Issuer but Subject follows the same format) that restricts there to be a single CN. The Subject field is just a list of type/values. There is a CAB thread on restricting issuance of certs with multiple CNs: https://cabforum.org/pipermail/public/2016-April/007242.html but I don't think that's a strong argument for being restrictive at the s2n server level.

I think we should support multiple CNs for the following reasons:
- If the application provides a server cert to s2n with multiple CNs, it makes sense to try to our best to support it. s2n will only consider the CNs if there are no valid SANs available.
- openssl/boring ssl walk over the CNs for their hostname matching function: https://github.com/google/boringssl/blob/c67076d653f7501136f7b208df4b011a7275e8f5/crypto/x509v3/v3_utl.c#L1066-L1077 https://github.com/openssl/openssl/blob/4286ca47c7fac2b5a6208b377967e0f095e9abc5/crypto/x509v3/v3_utl.c#L885-L896
- I was able to generate a certificate with multiple CNs. See https://github.com/awslabs/s2n/blob/72f90f6b0315499cb68376fb326d89d7214a7877/tests/pems/sni/quail_cn_rattlesnake_cn_cert.pem

Reply:
Cool. I agree with your rationale for multiple CNs and like the comment you've added